Docs: Recipe for Contentful data source (#17318)
gatsbyjs/gatsby: docs/docs/recipes.md

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing data from Contentful

#### Prerequisites

- A [Contentful account](https://www.contentful.com/)
- A Contentful space
- One [Content model](https://www.contentful.com/developers/docs/concepts/data-model/)
- One or more pieces of published content
- The [space ID and content delivery API key](https://www.contentful.com/developers/docs/concepts/apis/#content-delivery-api) for your space

For the screenshots and examples in this recipe, we have set up a new Contentful space with a "News" content model that includes "Title" and "Body" fields.

**Note:** Per [this issue](https://github.com/gatsbyjs/gatsby/issues/15344#issuecomment-520151776), you should create and publish at least one piece of media in your Contentful space to avoid build errors.

#### Directions

##### Install and setup the Contentful plugin

Install the `gatsby-source-contentful` plugin in your Gatsby site:

```shell
npm install --save gatsby-source-contentful
```

> Add the following plugin entry to your `gatsby-config.js` file. Use your own space ID and access token from the API keys section of your Contentful space. You may also use [environment variables](https://www.gatsbyjs.org/docs/environment-variables/) to store your space and token.

Reviewer: ```suggestion
2. Add the following plugin entry to your `gatsby-config.js` file. Use your own space ID and access token from the API keys section of your Contentful space. You may also use [environment variables](/docs/environment-variables/) to store your space and token.
```

```javascript:title=gatsby-config.js
plugins: [
  {
    resolve: `gatsby-source-contentful`,
    options: {

>       spaceId: `your_space_id`,

Reviewer: Should we add here directly the env to show good practice or it is out of scope here?

```suggestion
      spaceId: process.env.CONTENFUL_SPACE_ID,
```

Author: I went back and forth on that, I think by linking to the environment variable page we might be covered.

Reviewer 2: I'm sorta torn by this, too. Does committing it as-is present a security risk?

Reviewer: as I read from the docs there are readonly keys and also management access keys. If a newbie now expose this keys to a GitHub repo then it is not that good. Content can be manipulated or also (the user thinks it's) hidden fields can be read which are not for.public.

Reviewer 3: env vars might make this recipe harder, what if we put in a disclaimer about not exposing it to the public. Maybe we need a page on gatsby/docs that talks about secrets.

Reviewer: even the [Contentful README](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/README.md) shows this hint and a shortcut to https://gatsby.dev/env-vars

```
        spaceId: `your_space_id`,
        // Learn about environment variables: https://gatsby.dev/env-vars
        accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
```

Reviewer: i think we make a receipt not harder, when we show new users directly at the beginning the only secure way.

Author: Maybe we should reword the sentence before the example to say "You should" instead "You may"?

Also, blushing a bit here, I included a space and access token so that people could just use it against a known space - which I think might be a bit flimsy.

Reviewer: @wardpeet we have the https://gatsby.dev/env-vars page, but there is no section about security of private keys. good point.

Reviewer 2: Works for me!

Author: Added in https://github.com/gatsbyjs/gatsby/pull/17318/commits/1eb7a30ca348992fc0223370e63084c3cc9466a9, should we consider this resolved for now? I agree that a general best practice for secure variables should be outlined in https://gatsby.dev/env-vars - should an issue be opened for that?

Reviewer 2: That sounds like a great follow-up issue, so we can keep the scope of this PR focused.

>       accessToken: `your_access_token`,

Reviewer: ```suggestion
      accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
```

Author: Marking as resolved per https://github.com/gatsbyjs/gatsby/pull/17318#discussion_r320567189

    },
  },
]
```

Run `gatsby develop` and make sure the site compiled successfully.

##### Querying your data

Open the [Gatsby GraphiQL editor](https://www.gatsbyjs.org/docs/introducing-graphiql/) by visiting `https://localhost:8000/___graphql`. The Contentful plugin adds several new node types to your site, including a node type for every content type in your Contentful website. For our example site with a `News` content type and two fields, we get an `allContentfulNews` node type in GraphQL.

![the graphql interface, with a sample query outlined below](./images/recipe-sourcing-contentful-graphql.png)

To query all our news article from Contentful, we can use the following GraphQL query. Note that the `body` field is an object. [Long-text field types in Contentful are returned as objects](https://www.gatsbyjs.org/packages/gatsby-source-contentful/#a-note-about-longtext-fields) instead strings.

```graphql
{
  allContentfulNews {
    edges {
      node {
        body {
          body
        }
        title
      }
    }
  }
}
```

Contentful nodes also include several metadata fields like `createdAt` or `node_locale`.

##### Using Contentful data in your site

> You can use Contentful data like any other data source. A simple page listing news articles from our sample space would look like this:

Reviewer: Double space 

```suggestion
5. To source Contentful data in your Gatsby site, create a page component with a GraphQL query and some code to iterate over news articles from your sample space:
```

```jsx
import React from "react"
import { graphql } from "gatsby"

const NewsPage = ({ data }) => (
  <div>
    <h1>News</h1>
    {data.allContentfulNews.edges.map(({ node }) => (
      <>
        <h2>{node.title}</h2>
        <p>
          <strong>{node.createdAt}</strong>
          {" — "}
          {node.body.body}
        </p>
      </>
    ))}
  </div>
)

export default NewsPage

export const query = graphql`
  {
    allContentfulNews(sort: { fields: [updatedAt] }) {
      edges {
        node {
          title
          body {
            body
          }
          createdAt(formatString: "MMMM DD YYYY")
        }
      }
    }
  }
```

#### Additional resources

- [Building a Site with React and Contentful](https://www.gatsbyjs.org/blog/2018-1-25-building-a-site-with-react-and-contentful/)
- [More on Sourcing from Contentful](https://www.gatsbyjs.org/docs/sourcing-from-contentful/)
- [Contentful source plugin](https://www.gatsbyjs.org/packages/gatsby-source-contentful/)

##### Live example

> - [Example website](http://kevee.net/gatsby-source-contentful-recipe/)

Reviewer: Thx for contributing the demo. It is ok to have an example on an external server? Maybe we should host it on the Gatsby demo server?

Reviewer 2: Oh yeah - the convention is to include them in the Gatsby repo. Here's a recipe that does it: https://www.gatsbyjs.org/docs/recipes/#sourcing-markdown-data-for-blog-posts-and-pages-with-graphql

Author: I can include the example in the repo, but now that I look at it, there isn't anything in the example repo that isn't already outlined in the recipe itself. Should I add the example, or remove references to the live example?

Reviewer 2: I think we should add the example - it would help adoption and get people playing around with it! It makes sense for more in-depth recipes like this one - perhaps not for some of the smaller-scoped ones.

Author: Added example in https://github.com/gatsbyjs/gatsby/pull/17318/commits/9d341360d5eaa00ca341816a64551c6286a10c83. Also, the link in the recipe will of course be broken for now until the pull is merged.

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing data from Contentful

#### Prerequisites

- A [Contentful account](https://www.contentful.com/)
- A [Contentful space](https://www.contentful.com/faq/basics/#how-do-i-create-a-space)
- One [Content model](https://www.contentful.com/developers/docs/concepts/data-model/)
- One or more pieces of published content to [avoid build errors](https://github.com/gatsbyjs/gatsby/issues/15344#issuecomment-520151776)
- The [space ID and content delivery API key](https://www.contentful.com/developers/docs/concepts/apis/#content-delivery-api) for your space

> The screenshots and examples in this recipe were set up with a new Contentful space with a "News" content model that includes "Title" and "Body" fields.

Reviewer 3: When going through this recipe Title & Body don't have the same type.

`Title` is a short text type & `Body` is a long text type. I'm not sure if we should make this a bit more clear.

Author: Agreed, added in https://github.com/kevee/gatsby/commit/95facc99e25331a5e576db366590b56f62e7cb99

    },
  },
]
```

3. Run `gatsby develop` and make sure the site compiled successfully.

4) Query data with the [GraphiQL editor](/docs/introducing-graphiql/) at `https://localhost:8000/___graphql`. The Contentful plugin adds several new node types to your site, including every content type in your Contentful website. For example, a site with a `News` content type and two fields produces an `allContentfulNews` node type in GraphQL.

![the graphql interface, with a sample query outlined below](./images/recipe-sourcing-contentful-graphql.png)

To query for news articles from Contentful, use the following GraphQL query:

```graphql
{
  allContentfulNews {
    edges {
      node {
        body {
          body
        }
        title
      }
    }
  }
}
```

Contentful nodes also include several metadata fields like `createdAt` or `node_locale`.

5. To source Contentful data in your Gatsby site, create a page component with a GraphQL query and some code to iterate over news articles from your sample space:

```jsx
import React from "react"
import { graphql } from "gatsby"

const NewsPage = ({ data }) => (
  <div>

>     <h1>News</h1>

Reviewer 2: ```suggestion
    <h1>News</h1>
    <ul>
```

>     {data.allContentfulNews.edges.map(({ node }) => (

Reviewer 2: ```suggestion
    {data.allContentfulNews.edges.map(({ node, index }) => (
```

>       <>

Reviewer 2: ```suggestion
      <li key={index}>
```

        <h2>{node.title}</h2>
        <p>
          <strong>{node.createdAt}</strong>
          {" — "}

>           {node.body.body}

Reviewer 3: When setting body as a long text node, you get the option to make your text bold, italic, ...
![image](https://user-images.githubusercontent.com/1120926/65020211-04df2780-d92d-11e9-8fef-8e677e9532ad.png)

When just spitting out `body.body` and make a piece of the body bold.
![image](https://user-images.githubusercontent.com/1120926/65020283-2dffb800-d92d-11e9-841c-3ff3e919ca12.png)

Unsure if we need to put this in the recipe, but maybe some extra notes might help.

Author: Yes, a prior version of the recipe mentioned that, but the gatsby-source-contentful readme does a pretty good job covering that topic and more as well.

Author: I added a markdown transform to the recipe example to address this.

        </p>

>       </>

Reviewer 2: ```suggestion
      </li>
```

    ))}

>   </div>

Reviewer 2: ```suggestion
    </ul>
  </div>
```

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing data from Contentful

#### Prerequisites

> - A [Contentful account](https://www.contentful.com/)

Reviewer 2: ```suggestion
- A [Gatsby site](/docs/quick-start/)
- A [Contentful account](https://www.contentful.com/)
```

- The [Contentful CLI](https://www.npmjs.com/package/contentful-cli) installed

#### Directions

> 1. Login to Contentful with the CLI. It will help you create an account if you don't have one.

Reviewer 2: ```suggestion
1. Log in to Contentful with the CLI and follow the steps. It will help you create an account if you don't have one.
```

> ```

Reviewer: ````suggestion
```shell
````

contentful login
```

2. Create a new space. Make sure to save the space ID given to you at the end of the command.

> ```

Reviewer: ````suggestion
```shell
````

contentful space create --name 'Gatsby example'
```

> 3. Seed the new space with example blog content using the new space ID returned from the previous command.

Reviewer 2: ```suggestion
3. Seed the new space with example blog content using the new space ID returned from the previous command, in place of `<space ID>`.
```

The square brackets look like CLI syntax, so I could see people getting confused.

> ```

Reviewer: ````suggestion
```shell
````

> contentful space seed -s '[space ID]' -t blog

Reviewer 2: ```suggestion
contentful space seed -s '<space ID>' -t blog
```

> ```

Reviewer 2: ````suggestion
```

For example, with a space ID in place: `contentful space seed -s '22fzx88spbp7' -t blog`
````
I think it would help to show an example.

4. Create a new access token for your space.

> ```

Reviewer: ````suggestion
```shell
````

> contentful space accesstoken create -s '[space ID]' --name 'Example token'

Reviewer 2: ```suggestion
contentful space accesstoken create -s '<space ID>' --name 'Example token'
```

```

5. Install the `gatsby-source-contentful` plugin in your Gatsby site:

```shell
npm install --save gatsby-source-contentful
```

> 6. Edit the file `gatsby-config.js` to enable the `gatsby-source-contentful` plugin. You should consider using [environment variables](https://www.gatsbyjs.org/docs/environment-variables/) to store your space ID and token.

Reviewer 2: ```suggestion
6. Edit the file `gatsby-config.js` to enable the `gatsby-source-contentful` plugin. You should strongly consider using [environment variables](/docs/environment-variables/) to store your space ID and token for security purposes.
```

```javascript:title=gatsby-config.js
plugins: [
  {
    resolve: `gatsby-source-contentful`,
    options: {

>       spaceId: `[space ID]`, // or process.env.CONTENTFUL_SPACE_ID

Reviewer 2: ```suggestion
      spaceId: `<space ID>`, // or process.env.CONTENTFUL_SPACE_ID
```

>       accessToken: `[access token]`, // or process.env.CONTENTFUL_TOKEN

Reviewer 2: ```suggestion
      accessToken: `<access token>`, // or process.env.CONTENTFUL_TOKEN
```

    },
  },
]
```

7. Run `gatsby develop` and make sure the site compiled successfully.

8. Query data with the [GraphiQL editor](/docs/introducing-graphiql/) at `https://localhost:8000/___graphql`. The Contentful plugin adds several new node types to your site, including every content type in your Contentful website. Your example space with a "Blog Post" content type produces a `allContentfulBlogPost` node type in GraphQL.

![the graphql interface, with a sample query outlined below](./images/recipe-sourcing-contentful-graphql.png)

> To query for a Blog Posts titles from Contentful, use the following GraphQL query:

Reviewer 2: ```suggestion
To query for Blog Post titles from Contentful, use the following GraphQL query:
```

```graphql
{
  allContentfulBlogPost {
    edges {
      node {
        title
      }
    }
  }
}
```

Contentful nodes also include several metadata fields like `createdAt` or `node_locale`.

9. To show a simple list of blog posts, create a new file in `/src/pages/blog.js`. This page will display all posts, sorted by updated date.

```jsx:title=src/pages/blog.js
import React from "react"
import { graphql, Link } from "gatsby"

const BlogPage = ({ data }) => (
  <div>
    <h1>Blog</h1>
    <ul>
      {data.allContentfulBlogPost.edges.map(({ node, index }) => (
        <li key={index}>
          <Link to={`/blog/${node.slug}`}>{node.title}</Link>
        </li>
      ))}
    </ul>
  </div>
)

export default BlogPage

export const query = graphql`
  {
    allContentfulBlogPost(sort: { fields: [updatedAt] }) {
      edges {
        node {
          title
          slug
        }
      }
    }
  }
`
```

> 

Reviewer 2: ```suggestion

To continue building out your Contentful site including post detail pages, check out the rest of the [Gatsby docs](/docs/sourcing-from-contentful/) and additional resources below.
```

To query for Blog Post titles from Contentful, use the following GraphQL query:

```graphql
{
  allContentfulBlogPost {
    edges {
      node {
        title
      }
    }
  }
}
```

Contentful nodes also include several metadata fields like `createdAt` or `node_locale`.

9. To show a simple list of blog posts, create a new file in `/src/pages/blog.js`. This page will display all posts, sorted by updated date.

```jsx:title=src/pages/blog.js
import React from "react"
import { graphql, Link } from "gatsby"

const BlogPage = ({ data }) => (
  <div>
    <h1>Blog</h1>
    <ul>
      {data.allContentfulBlogPost.edges.map(({ node, index }) => (
        <li key={index}>
          <Link to={`/blog/${node.slug}`}>{node.title}</Link>
        </li>
      ))}
    </ul>
  </div>
)

export default BlogPage

export const query = graphql`
  {
    allContentfulBlogPost(sort: { fields: [updatedAt] }) {
      edges {
        node {
          title
          slug
        }
      }
    }
  }
`
```

To continue building out your Contentful site including post detail pages, check out the rest of the [Gatsby docs](/docs/sourcing-from-contentful/) and additional resources below.

> #### Additional resources

Reviewer: ```suggestion

#### Additional resources
```

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing data from Contentful

#### Prerequisites

- A [Gatsby site](/docs/quick-start/)
- A [Contentful account](https://www.contentful.com/)
- The [Contentful CLI](https://www.npmjs.com/package/contentful-cli) installed

#### Directions

> 1. Log in to Contentful with the CLI and follow the steps. It will help you create an account if you don't have one.

Reviewer 2: @muescha recipes are intentionally concise; we should limit this to what is actionable.

> 1. Log in to Contentful with the CLI and follow the steps. It will help you create an account if you don't have one.

Reviewer: yeah - but i was asking me this by myself. and if my commandline gets messy and cleared somehow i don't know if i need it later and copy it somewhere before i go on. so a short info like "the token you get here is only for connection the cli. you don't need it later."

> 1. Log in to Contentful with the CLI and follow the steps. It will help you create an account if you don't have one.

Reviewer: with out this info i start googling and get distracted before i go on with the receipe

```shell
contentful login
```

> 2. Create a new space. Make sure to save the space ID given to you at the end of the command.

Reviewer 2: ```suggestion
2. Create a new space if you don't already have one. Make sure to save the space ID given to you at the end of the command. If you already have a Contentful space and space ID, you can skip this step.

Note: for new accounts, you can overwrite the default onboarding space. Check to see the [spaces included with your account](https://app.contentful.com/account/profile/space_memberships).
```

How about this, @muescha?

Reviewer: using the exisiting one can be a different template and not the blog template and then the other code in this receipe not works.

maybe just mention here that they have 2 free spaces...

or only go on when it is a blog template. but if we use a existing space, then it gets tricky because the user need to find out where he find the tokens.

```shell
contentful space create --name 'Gatsby example'
```

3. Seed the new space with example blog content using the new space ID returned from the previous command, in place of `<space ID>`.

```shell
contentful space seed -s '<space ID>' -t blog
```

For example, with a space ID in place: `contentful space seed -s '22fzx88spbp7' -t blog`

4. Create a new access token for your space.

```shell
contentful space accesstoken create -s '<space ID>' --name 'Example token'
```

5. Install the `gatsby-source-contentful` plugin in your Gatsby site:

```shell
npm install --save gatsby-source-contentful
```

6. Edit the file `gatsby-config.js` to enable the `gatsby-source-contentful` plugin. You should strongly consider using [environment variables](/docs/environment-variables/) to store your space ID and token for security purposes.

```javascript:title=gatsby-config.js
plugins: [

>   {

Reviewer 2: I don't think this comment is as effective as it could be, as it appears related to the following code instead of outside of it. I'd recommend something like this:

```suggestion
   // add to array along with any other installed plugins
   // highlight-start
   {

```

>   {

Reviewer: yeh thats better

    resolve: `gatsby-source-contentful`,
    options: {
      spaceId: `<space ID>`, // or process.env.CONTENTFUL_SPACE_ID
      accessToken: `<access token>`, // or process.env.CONTENTFUL_TOKEN
    },
  },
]
```

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer 2: You shouldn't need this -- sounds like an issue with your Gatsby CLI installation. Creating a Gatsby site with the Quick Start should install the other required dependencies. I'd recommend opening a new issue for that and limiting the scope of the recipe.

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer: i replayed the example from beginning. done all the commands, also the `npm install -g gatsby-cli` and get at this point an error.

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer: the `npm install` gives me this output and i see there are new packages installed and changed:

```bash
npm install
                                                                                                                                                                                                                                                                                                     
> gatsby@2.15.28 postinstall /Users/muescha/Work/gatsby/gatsby-site-contentful2/node_modules/gatsby                                                                                                                                                                                                  
> node scripts/postinstall.js                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                     
npm WARN eslint-config-react-app@4.0.1 requires a peer of eslint-plugin-flowtype@2.x but none is installed. You must install peer dependencies yourself.                                                                                                                                             
npm WARN tsutils@3.17.1 requires a peer of typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta but none is installed. You must install peer dependencies yourself.                                  
                                                                                                                                                                                                                                                                                                     
added 5 packages from 2 contributors, removed 5 packages, updated 13 packages and audited 28684 packages in 17.208s                                                                                                                                                                                  
found 0 vulnerabilities 
```

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer: install path:

```bash
npm install -g gatsby-cli
gatsby new gatsby-site-contentful2
cd gatsby-site-contentful2/
gatsby develop
gatsby build
gatsby serve
# in this case leave off the contentful cli commands
npm install --save gatsby-source-contentful
nano gatsby-config.js  # add tokens
gatsby develop         # error
npm install
gatsby develop
```

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer: opened new issue: #18048

> 7. Run `gatsby develop` and make sure the site compiled successfully.

Reviewer: i mark this conversation as resolved

8. Query data with the [GraphiQL editor](/docs/introducing-graphiql/) at `https://localhost:8000/___graphql`. The Contentful plugin adds several new node types to your site, including every content type in your Contentful website. Your example space with a "Blog Post" content type produces a `allContentfulBlogPost` node type in GraphQL.

![the graphql interface, with a sample query outlined below](./images/recipe-sourcing-contentful-graphql.png)

To query for Blog Post titles from Contentful, use the following GraphQL query:

```graphql
{
  allContentfulBlogPost {
    edges {
      node {
        title
      }
    }
  }
}
```

> Contentful nodes also include several metadata fields like `createdAt` or `node_locale`.

Reviewer: ```suggestion
Contentful nodes also include several metadata fields like `createdAt` or `node_locale`. You select available fields in the left `Explorer` sidebar. Click on button `Explorer` to show this sidebar.
```

Reviewer 2: Recipes are supposed to be concise, so I'd recommend against including this here.

> 9. To show a simple list of blog posts, create a new file in `/src/pages/blog.js`. This page will display all posts, sorted by updated date.

Reviewer: ```suggestion
9. To show a list of links to the blog posts, create a new file in `/src/pages/blog.js`. This page will display all posts, sorted by updated date.
```

avoid simple and easy also here?

```jsx:title=src/pages/blog.js
import React from "react"
import { graphql, Link } from "gatsby"

const BlogPage = ({ data }) => (
  <div>
    <h1>Blog</h1>
    <ul>
      {data.allContentfulBlogPost.edges.map(({ node, index }) => (
        <li key={index}>
          <Link to={`/blog/${node.slug}`}>{node.title}</Link>
        </li>
      ))}
    </ul>
  </div>
)

export default BlogPage

export const query = graphql`
  {
    allContentfulBlogPost(sort: { fields: [updatedAt] }) {
      edges {
        node {
          title
          slug
        }
      }
    }
  }
`
```

> To continue building out your Contentful site including post detail pages, check out the rest of the [Gatsby docs](/docs/sourcing-from-contentful/) and additional resources below.

Reviewer: i missed a bonus -> the created page in browser :)

```suggestion
10. Open your browser and navigate to the new page: `http://localhost:8000/blog`. You see a list of links to the blog articles. If you click on a link you get the Gatsby 404 error page until you created the blog pages from a template (See tutorial/guide xxx)
(If the Gatsby development server is not running anymore restart it with `gatsby develop`.)

To continue building out your Contentful site including post detail pages, check out the rest of the [Gatsby docs](/docs/sourcing-from-contentful/) and additional resources below.
```

and i miss also the generated blog pages from contentful. maybe show this also here?

Reviewer 2: Again -- recipes should be concise. Given this already has 10 steps, I would recommend limiting the scope and linking off for additional resources.

Reviewer: ok - thats fine to wrap up the receipte to the result and then point to the next guide
